### PR TITLE
Support external imports in functions

### PR DESCRIPTION
**Pull request by jdahm**
## Description

Support external imports in functions uniformly. These were already supported if the functions were called inline, however if the functions were an external, they were not resolved.

This contains a small fix for the frontend, as well as a terribly pathological test.

## Requirements

Before submitting this PR, please make sure:

- [x] The code builds cleanly without new errors or warnings
- [x] The code passes all the existing tests
- [x] If this PR adds a new feature, new tests have been added to test these
new features
- [x] All relevant documentation has been updated or added

Additionally, if this PR contains code authored by new contributors:

- [x] All the authors are covered by a valid contributor assignment agreement,
signed by the employer if needed, provided to ETH Zurich
- [x] The names of all the new contributors have been added to an updated
version of the AUTHORS.rst file included in the PR

### CODE REVIEW

File: src/gt4py/frontend/gtscript_frontend.py [hunk 1]
```diff
@@ -1468,6 +1466,33 @@ def resolve_external_symbols(
...
+
+                # Get sub-context
+                sub_context, sub_unbound = gt_meta.get_closure(
+                    value, included_nonlocals=True, include_builtins=False
+                )
+
+                # Resolve imports or add to list to recursively resolve
+                for imported_name, imported_value in value._gtscript_["imported"].items():
+                    if isinstance(imported_value, types.FunctionType):
+                        add_func_externals[imported_name] = sub_context[imported_value]
```
Comment (egparedes) [inline src/gt4py/frontend/gtscript_frontend.py:1490]: I guess this should be:

```suggestion
                        new_func_externals[imported_name] = sub_context[imported_value]
```

If the tests are passing even with this typo in the code, there is something else wrong either in the code or in the tests.

**Review by egparedes — CHANGES_REQUESTED** [review]
Thanks for finding and fixing this bug. The PR looks fine but there is a typo in the code that should make the test fail. Please fix the typo and verify that the added tests is well defined and covers all the corner cases.

File: src/gt4py/frontend/gtscript_frontend.py [hunk 2]
```diff
@@ -1468,6 +1466,33 @@ def resolve_external_symbols(
         resolved_imports = {**imported}
         resolved_values_list = list(nonlocals.items())
 
+        # Resolve function-like imports recursively
+        func_externals = dict(
+            filter(
+                lambda name_value: isinstance(name_value[1], types.FunctionType), context.items()
+            )
+        )
```
[the remark this answers is not in the document]
Comment (jdahm) [reply src/gt4py/frontend/gtscript_frontend.py:1474]: I keep forgetting that comprehensions can have if statements on them! Thanks for the suggestion.

**Comment by jdahm** [comment]
GitHub Actions failed here. There is no issue with formatting. I'll try to push a minor commit so that it rebuilds after ensuring there is full test coverage.

File: tests/test_unittest/test_gtscript_frontend.py [hunk 3]
```diff
@@ -175,6 +175,64 @@ def definition_func(inout_field: gtscript.Field[float]):
...
+        stencil_id, def_ir = compile_definition(
+            definition_func,
+            "test_recursive_imports",
+            module,
+            externals={"some_call": some_function, "const": GLOBAL_CONSTANT},
+        )
+        assert set(def_ir.externals.keys()) == {
+            "some_call",
+            "const",
+        }
```
Comment (jdahm) [inline tests/test_unittest/test_gtscript_frontend.py:201]: For reasons that still elude me, this test is necessary. I thought this was covered below, but this uncovered the need for the additional annotate call at gtscript_frontend.py:1477.

File: tests/test_unittest/test_gtscript_frontend.py [hunk 4]
```diff
@@ -175,6 +175,64 @@ def definition_func(inout_field: gtscript.Field[float]):
...
+            "test_recursive_imports",
+            module,
+            externals={"some_call": func_nest1, "other": func_nest2, "const": GLOBAL_CONSTANT},
+        )
+        assert set(def_ir.externals.keys()) == {
+            "some_call",
+            "const",
+            "other",
+            "func_nest1",
+            "tests.test_unittest.test_gtscript_frontend.func_nest1.func_nest2",
```
Comment (jdahm) [inline tests/test_unittest/test_gtscript_frontend.py:233]: Externals captured implicitly get the fully qualified name.

**Review by egparedes — APPROVED** [review]
The code looks correct although it's hard to say if all the corner cases are covered. Probably the cause of all the weird cases is allowing calls to undecorated functions, which I still think feels like the wrong pattern. Anyway, if you think this PR is done, we can merge it.

**Comment by jdahm** [comment]
Yes, I believe this PR is ready, but should have a follow-up that cleans up the annotation system.
